### PR TITLE
Unified integration

### PR DESCRIPTION
**Pull request by mlamarre**
This is the port of Joan Alabort CVPR 2015 code to the master menpofit repository as we discussed on googlegroups. 

I kept Joan's history for files copied over. On the command line you need to use git log --follow to see the changes prior to the repo to repo copy. 

I tested with conda python 3.5 and 2.7 on windows. 

My current test program is here:

https://github.com/VLAM3D/alabortcvpr2015/blob/just_test/test/train_aic_rlms.py

but there's nothing special about it.

### CODE REVIEW

**Comment by jalabort** [comment]
Haven't really checked it yet, but my first impression by just skimming through the code is that it looks very neat :+1. Thanks for taking on this work @mlamarre!

**Comment by patricksnape** [comment]
Wow thanks a lot for this work @mlamarre ! Very cool to have someone contribute this much work! Haven't run your script yet but just wondering - did you replicate @jalabort results with this?

**Comment by patricksnape** [comment]
jenkins test this please

**Comment by mlamarre** [comment]
How do you run the tests locally?

The details show an error when uploading to anaconda.org

binstar_client.errors.NotFound: ('Not Found: Nothing matches the given URI ([GET] https://api.anaconda.org/channels/menpo/pr/94/merge -> 404)', 404)

**Comment by patricksnape** [comment]
Yes that is expected at the moment, we are currently handling a bug due to a security fix on our jenkins build box for OSX builds.

**Comment by patricksnape** [comment]
jenkins retest this please

**Comment by patricksnape** [comment]
jenkins retest this please

**Comment by mlamarre** [comment]
I didn't know that "jenkins test this please" was an invocation for the build system :)

I didn't try to replicate Joan's paper result, I refactored while checking a small regression dataset making notes on the differences and using the PTVS debugger to output intermediate variables to identify the changes. The regression dataset is obtained by training on the first 100 LFPW and testing on the first 16 AFW. 

Two sources of difference from the CVPR 2015 repo code:
- the now correct subpixel transforms;
- the mmcf experts computes different filters, extract_patches() is slightly different, with another subpixel fix there, but even for the same input the learned filter differs (the desired response PDF are identical).

The new expert code seems more sophisticated and more correct with a context_patch size to deal with boundaries. 

So the results changed noticeably, the differences are random but correlated as we can see below: 

image

The raw data is there (one csv file per change of results):

https://github.com/VLAM3D/alabortcvpr2015/tree/just_test

I didn't change anything of importance to the algorithm run() methods, I might have set un-optimal default values for the multiple parameters because they are just based on the latest Nov 2014 commit. They may not be the same as the paper parameters.

Note 1: I see numpy warnings in the scope of menpofit.builder.build_reference_frame(). The only input is the shape model mean. In my test program I silenced them. I see these warnings for the 66 and 68 TriMesh with any subset of the LFPW training set. 

Note 2: Training the experts with holistic_features = fast_dsift is ~10x slower than using no_op. I don't know why. I verified that this occurs with plain CLM. Currently, in the Pull Request version the holistic_features arguments is shared by the AAM and CLM. Maybe it would be better to have distinct settings. Your call. 

I will certainly give menpobench a try, but I was not planning on trying to get optimal results on "in the wild images". In my application, people don't smoke cigars, wear glasses or have hands over their faces. 

I will now focus on tracking and probably trying to use particle filter concepts in the fitting algorithm.

**Comment by jalabort** [comment]
Hi Mathieu,

The current `mmcf` experts in `menpofit` are expected to work better than the preliminary version on my old repo. This should not be a concern.

`fast_dsift` take some time to compute (depending on the size of your input images), `no_op` is faster because it does not compute anything. Moreover, ` fast_dsift` changes the 3 channel (RGB) representation of the original images for one that is based on reduced SIFT descriptors which is made up of 8 channels.  Therefore, when using `fast_dsift`, the dimensionality of the patches used to train the experts will be MxNx8  instead of MxNx3 (where MxN refers to the patch size, e.g. 17x17). This increase in dimensionality will effect the time required to learn the each expert (and, typically, there are 60 something of them). Hopefully this helps explain why it takes longer to train using `fast_dsift`. 

I would say it would be better to have two separate `holistic_features` arguments. It is an easy change and would make the algorithm much more flexible. However, the core menpofit guys @patricksnape @nontas should certainly comment on this.

By the look of your graph I’d say that you code is probably correct but it wouldn’t hurt running a complete experiment on LFPW comparing it agains basic `AIC` and `RLMS`.

**Comment by patricksnape** [comment]
Ok just doing a code review and there are a few issues - I'll put a PR up against your PR @mlamarre to save you the trouble.

**Comment by patricksnape** [comment]
Ok that PR is [here](https://github.com/VLAM3D/menpofit/pull/1). And a notebook to do some testing with is [here](https://gist.github.com/patricksnape/544c66b20ae5eb20f9632e5ed71ee4aa).

**Comment by patricksnape** [comment]
Results for the notebook linked to, using uniform noise perturbations of the ground truth bounding box with std 0.05. Average edge length normalised by 68 points. Trained/tested on LFPW.

screen shot 2016-05-26 at 23 33 10

**Comment by jalabort** [comment]
Nice graph @patricksnape. Do you have the corresponding `GlobalAAM` and `CLM` graphs for this experiments (with more or less the same parameters)?

**Comment by patricksnape** [comment]
Don't worry about it - anaconda is just overloaded hence the failures.

**Comment by mlamarre** [comment]
I have been running tests since the last activity on this thread. I have several pickled models trained on ~3k images for various algorithms (HAAM, PAAM, RegularizedSDM, CLM, Unified AIC-RLMS). I could make a CED on the LFPW testset, which is not part of my training list, but first I need to address a few issues.

I can't reproduce the same initial error curve as Patrick. The snippet I'm using to perturb the bb:
`bb = noisy_shape_from_bounding_box(gt_s.bounding_box(),gt_s.bounding_box())`

Which gives a much flatter (difficult) Initial curve. 

Also, do you have an updated CLM example notebook? The default values or the values from "CLMs Basic" give pretty bad results. CLM is the only algorithm giving me trouble. I saw an old answer in the mailing list that CLMs only work with some features (sparse_hog was suggested).  I run out of memory, 32 GB, with sparse_hog, I'm trying float32_fast_dsift right now.

Thanks

**Comment by mlamarre** [comment]
image

On the first 600 frames of sequence 115 in 300VW, the initial shape of a frame is the final shape from the previous. 

Compared to DlibWrapper with the public shape_predictor_68_face_landmarks.dat. DlibWrapper initial shape is the Dlib face detector bounding box for all frames, it performs better than using the previous frame final shape bounding box. 

I put a side-by-side in an unlisted YouTube. 

https://youtu.be/MfBQ1Q_zzPc

**Comment by jalabort** [comment]
Nice! I'm quite surprised by your tracking results! Is that measuring the error on all landmarks or only the inner ones?

Regarding the initialization for an LFPW @patricksnape is probably the best person to answer that since I'm not up to date with the very latest changes they introduced in menpofit.

Regarding your issues with memory. You can now use an incremental method for training CLMs https://github.com/menpo/menpofit/blob/master/menpofit/clm/base.py#L277 You can set batch size to a low number for which the memory issues should disappear. However, if you are having memory issues in CLMs you should also have them with Unified I believe... Is the size of the patches the same in both methods? Maybe @patricksnape or @nontas have more info regarding this... Am I missing something?

The answer on the mailing list is an old one and that problem should now be completely resolved. CLMs can be used with all the `spatial` features in menpo.

**Comment by patricksnape** [comment]
results

**Comment by jalabort** [comment]
Awesome @patricksnape. This is exactly how it should look like!  Based on this, I'm happy for this to be merged in when you guys are.

@mlamarre if you still having problems with `CLM` maybe it'd be a good idea to post a question on our mailing list to discuss that, so that this discussion is left clean.

Many thanks for bringing this to the new `menpofit` @mlamarre, awesome job!

File: menpofit/unified/algorithm.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+
+    r"""
+    Base interface for optimization of UnifiedAAMCLM
+
+    Parameters
+    ----------
+    aam_interface : : `subclass` of :map:`LucasKanadeBaseInterface`, 
+        Concrete instantiation of an interface for Lucas-Kanade optimization of
+        standard AAMs.
+    appearance_model : `menpo.model.PCAModel` or subclass
```
Comment (nontas) [inline menpofit/unified/algorithm.py]: The doc needs to be updated here. `appearance_model` and `transform` are not existent.

File: menpofit/unified/algorithm.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        initial_shape : `menpo.shape.PointCloud`
+            The initial shape from which the optimization will start.
+        gt_shape : `menpo.shape.PointCloud` or ``None``, optional
+            The ground truth shape of the image. It is only needed in order
+            to get passed in the optimization result object, which has the
+            ability to compute the fitting error.
+        max_iters : `int`, optional
+            The maximum number of iterations. Note that the algorithm may
+            converge, and thus stop, earlier.
+        prior : `bool`, optional
```
Comment (nontas) [inline menpofit/unified/algorithm.py]: `return_costs` needs to be added in the doc. Something like:
```
return_costs : `bool`, optional
    If ``True``, then the cost function values will be computed
    during the fitting procedure. Then these cost values will be
    assigned to the returned `fitting_result`. Note that the costs
    computation increases the computational cost of the fitting. The
    additional computation cost depends on the fitting method. Only
    use this option for research purposes.
```

File: menpofit/unified/algorithm.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
```
Comment (nontas) [inline menpofit/unified/algorithm.py]: Same here for `return_costs` in the doc

**Comment by nontas** [comment]
Amazing job @mlamarre !! The code is clean and seems to work, which is great!

I made 2-3 minor comments about some omissions in the doc strings. I also have the following comments:

1. We should consider returning a `UnifiedAAMCLMResult` instead of `ParametricIterativeResult`, in order to also return the `appearance_parameters`. This is exactly what we are currently doing for the AAM (https://github.com/menpo/menpofit/blob/master/menpofit/aam/result.py). What do you think @patricksnape , @jalabort ?

2. The model needs to be added in the API documentation. This is done by creating its own folder in `menpofit/docs/source/api/menpofit/` and create a `.rst` file for each class that is exposed along with an `index.rst`, exactly as it is done for the AAM (https://github.com/menpo/menpofit/tree/master/docs/source/api/menpofit/aam). 

3. I have some doubts about the selected names. I believe that `unified` is too general and something like `unified_aam_clm` should be used as module name instead. Also, `PICRLMS` and `AICRLMS` are not consistent with the way that we name the AAM and CLM algorithms. We are normally using full names, such as `ProjectOutInverseCompositional` and `RegularisedLandmarkMeanShift`. However, I do understand that `ProjectOutInverseCompositional_RegularisedLandmarkMeanShift` is way too long. Any thoughts @patricksnape  @jalabort ? 

Let me know if you want me to actually make the changes I mentioned on 1 and 2! Once again, thanks a lot for the contribution @mlamarre !

**Comment by jabooth** [comment]
@mlamarre just to reiterate - this contribution is amazing, thank you so much!

@nontas on the naming points, obviously it's ultimately your guys call, but I would raise the following in terms of wider consistency with menpo:

1. Try and avoid underscores at all costs in module names. This is particularly tricky though, as I agree `unified` is broad (unified what?) but `unifiedaamclm` is pretty illegible. Maybe this is the exception that proves the rule and `united_aam_clm` is the best of a bad bunch?
2. The same goes for class names `WhichReallyShouldBeOnlyCamelCaseWithNoUnderscores`. I would rather `ProjectOutInverseCompositionalRegularisedLandmarkMeanShift` than `ProjectOutInverseCompositional_RegularisedLandmarkMeanShift` - but if you go with the underscore for the module we could have a pattern of underscore denoting combined methods?

As always the main thing is consistency, so as long as there is a rule decided for this case which can be followed in similar cases down the line I have no issues whatever you decide!

**Comment by patricksnape** [comment]
`ProjectOutInverseCompositionalRegularisedLandmarkMeanShift` is madness. It's difficult to read it's so long. But it would be inconsistent to rename it to anything other than that. Still, maybe for the sake of this method we allow shorthands and just explain in the docs what they are? Or maybe just shorten `InverseCompositional -> IC` since we are unlikely to implement anything else?

**Comment by nontas** [comment]
Actually we could totally omit `IC`. So I vote for `unified_aam_clm`, `ProjectOutRegularisedLandmarkMeanShift` and `AlternatingRegularisedLandmarkMeanShift`.

**Comment by jabooth** [comment]
> `ProjectOutInverseCompositionalRegularisedLandmarkMeanShift` is madness.

@patricksnape pfft that is but a [wee bairn](https://sco.wikipedia.org/wiki/Bairn) of a name laddy: https://github.com/Quotation/LongestCocoa

But seriously, if the name is over specified could we just get by without the `InverseCompositional` altogether? (`ProjectOutRegularisedLandmarkMeanShift`)

**Comment by nontas** [comment]
Great minds think alike @jabooth :smile:

**Comment by jalabort** [comment]
Yeah I'm with @ and @jabooth getting rid of `IC` seems appropriate here.

@jabooth what on earth is a wee bairn? That link made no sense to me

**Comment by jabooth** [comment]
@nontas apparently so mate! ;)

@jalabort haha it's used in Scotland colloquially, means 'child':
http://www.urbandictionary.com/define.php?term=bairn

**Comment by mlamarre** [comment]
The credit for this PR goes to the Menpo team, this is all your code. I'm just a beta tester here.
Also thanks for the conda channel, you're now my first provider of binaries, replacing Christoph Gohlke. 

@jalabort the errors in the time series are what I get from fr.final_result() with 68 landmarks

If you want me to do name changes, I can do it, I'll wait for your final decision.

**Comment by patricksnape** [comment]
:+1:  Agree with @nontas on naming

**Comment by nontas** [comment]
Great! So @mlamarre just change the names to `unified_aam_clm`, `ProjectOutRegularisedLandmarkMeanShift` and `AlternatingRegularisedLandmarkMeanShift` and this PR should be ready. 

Let's keep the returned fitting result object as it is for now and I'll make a separate PR to create a model specific result, define `view_*_widget()` methods and add the model in the `docs`.

**Comment by patricksnape** [comment]
Ignore appveyor.

Jenkins test this please.

**Comment by patricksnape** [comment]
Jenkins test this please

**Comment by nontas** [comment]
In order to merge this before there are too many changes in the master that make merging impossible, I create a PR to @mlamarre's branch [**here**](https://github.com/VLAM3D/menpofit/pull/2).

The PR:
* Merges the current master.
* Adds `UnifiedAAMCLMAlgorithmResult` and `UnifiedAAMCLMResult` in `menpofit.unified_aam_clm.result`. The internal loop of the fitting algorithms is changed as part of this, in order to follow the logic of `menpofit`'s AAM and CLM.
* Various `view_*_widget()` methods are added in `UnifiedAAMCLM` for the shape model, appearance model, AAM, experts ensemble and CLM.
* Adds the model to the docs and README.md.

I ran an experiment with the following parameters:
```
diagonal =150
scales = (0.25, 0.5, 1.0)
holistic_features = fast_dsift
max_shape_components = 15
max_appearance_components = 150
patch_shape = (17, 17)
context_shape = (34, 34)

n_shape = [5, 10, 15]
n_appearance = 150
max_iters = 30
```
using 
* `UnifiedAAMCLM`
* `AAM` (with `WibergInverseCompositional` and `ProjectOutInverseCompositional`)
* `CLM` (with `RegularisedLandmarkMeanShift`)

All models are trained on LFPW trainset and tested on LFPW testset. They are all initialised using the groundtruth shape's bounding box. The result using `menpofit.error.human.bb_avg_edge_length_68_euclidean_error` is:
result

Once @mlamarre accepts the PR, this should be ready.

**Comment by patricksnape** [comment]
@nontas That is a lot of code! Will have to wait for me to review later as a bit busy at the moment...

**Comment by nontas** [comment]
@patricksnape should we merge that?

**Comment by patricksnape** [comment]
yes

**Pull request merged**